[FSDP] summon offload to CPU

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73904
* #73903

Implement ability to offload full params to CPU in summon_full_params.

Differential Revision: [D34707801](https://our.internmc.facebook.com/intern/diff/D34707801/)

rohan-varma committed:
[FSDP] summon offload to CPU

Implement ability to offload full params to CPU in summon_full_params.

Differential Revision: [D34707801](https://our.internmc.facebook.com/intern/diff/D34707801/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] summon offload to CPU"


Implement ability to offload full params to CPU in summon_full_params.

Differential Revision: [D34707801](https://our.internmc.facebook.com/intern/diff/D34707801/)

[ghstack-poisoned]

fegin reviewed (approved):
LGTM, one comment regarding the warning of the combination of rank_0 + cpu_offload

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1155,9 +1159,16 @@ def summon_full_params(
+            offload_to_cpu (bool, optional): If ``True``, full parameters are
+                offloaded to CPU. Note that this offloading currently only
+                occurs if the parameter is sharded (which is only not the case
+                for world_size = 1).
```
sounds good, added a warning

rohan-varma committed:
Update on "[FSDP] summon offload to CPU"


Implement ability to offload full params to CPU in summon_full_params.

Differential Revision: [D34707801](https://our.internmc.facebook.com/intern/diff/D34707801/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1231,6 +1261,19 @@ def _free_full_params_and_use_local_shard(params_to_free):
+                            )
+                            self._update_p_data(
+                                p, output_tensor=p._full_param_padded,  # type: ignore[attr-defined]
+                            )
```
@zhaojuanmao just want to double check this offloading approach since it's a bit different than regular parameter offload.

Here we have full_param_padded materialized, and p.data points to it. So to offload, we:

1. offload full param padded
2. Update p.data to point to the offloaded full param padded. 

And we restore appropriately when exiting the context.

If we directly offloaded p.data, it would still hold onto GPU memory because of full param padded.

zhaojuanmao replied:
yeah, this looks good to me!

zhaojuanmao reviewed (commented):
left some minor comments, looks great!

in terms of enabling these configs in state_dict() APIs, could we make rank0_only as default setting and cpu_offloading as optional?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1208,6 +1233,17 @@ def _free_full_params_and_use_local_shard(params_to_free):
+                                        )
+
                         if writeback:
                             self._write_back_current_shard()
```
I'll look into this and file an issue.

rohan-varma replied:
https://github.com/pytorch/pytorch/issues/74166

rohan-varma committed:
Update on "[FSDP] summon offload to CPU"


Implement ability to offload full params to CPU in summon_full_params.

Differential Revision: [D34707801](https://our.internmc.facebook.com/intern/diff/D34707801/)

[ghstack-poisoned]
